src/robusta/utils/cluster_provider_discovery.py

```diff
@@ -0,0 +1,120 @@
+    ClusterProviderType.Kops: "kops.k8s.io/instancegroup",
+    ClusterProviderType.Kapsule: "k8s.scaleway.com/kapsule"
+}
+
+__CLUSTER_PROVIDER__ = None
+
+
+def get_cluster_provider():
+    global __CLUSTER_PROVIDER__
+    return __CLUSTER_PROVIDER__
+
+
+def discover_cluster_provider():
+    global __CLUSTER_PROVIDER__
+    if not __CLUSTER_PROVIDER__:
+        __CLUSTER_PROVIDER__ = _find_cluster_provider().value
+        logging.info(f"{__CLUSTER_PROVIDER__} cluster detected")
+
+
+def _get_node_label(node, label) -> Optional[str]:
+    try:
+        return node.metadata.labels[label]
+    except KeyError:
+        pass
+    return None
+
+
+def _detect_provider_from_hostname(nodes) -> Optional[ClusterProviderType]:
+    nodes_host_names = [_get_node_label(node, "kubernetes.io/hostname") for node in nodes]
+    for host_name in nodes_host_names:
+        if not host_name:
+            continue
+        for cluster_type in HOSTNAME_MATCH:
+            cluster_hostname_regex = HOSTNAME_MATCH[cluster_type]
+            if len(re.findall(cluster_hostname_regex, host_name)) >= 1:
+                return cluster_type
+    return ClusterProviderType.Unknown
+
+
+def _detect_provider_from_node_labels(nodes) -> Optional[ClusterProviderType]:
+    for cluster_type in NODE_LABELS:
+        if _get_node_label(nodes[0], NODE_LABELS[cluster_type]):
+            return cluster_type
+    return ClusterProviderType.Unknown
+
+
+def _is_aks(nodes) -> bool:
+    try:
+        node = nodes[0]
+        provider_id = node.spec.providerID
+        if "aks" in provider_id:
+            return True
+    except (AttributeError, TypeError):
+        # is not aks, field is optional so could be missing
+        pass
+    return False
+
+
+def _is_detect_cluster_from_kubelet_version(nodes, kubelet_substring) -> bool:
+    try:
+        node = nodes[0]
+        kubelet_version = node.status.nodeInfo.kubeletVersion
+        if kubelet_substring in kubelet_version:
+            return True
+    except (AttributeError, TypeError):
+        # missing kubeletVersion
+        pass
+    return False
+
+
+def _find_cluster_provider() -> ClusterProviderType:
+    try:
+        nodes = NodeList.listNode().obj.items
+        cluster_hostname_provider = _detect_provider_from_hostname(nodes)
+        if cluster_hostname_provider != ClusterProviderType.Unknown:
+            return cluster_hostname_provider
+        elif _is_aks(nodes):
+            return ClusterProviderType.AKS
+        elif _is_detect_cluster_from_kubelet_version(nodes, 'gke'):
+            return ClusterProviderType.GKE
+        elif _is_detect_cluster_from_kubelet_version(nodes, 'eks'):
+            return ClusterProviderType.EKS
+
+        cluster_provider_from_labels = _detect_provider_from_node_labels(nodes)
+        if cluster_provider_from_labels != ClusterProviderType.Unknown:
+            return cluster_provider_from_labels
+
+    except Exception:
```

Comment:
It is actually bad to have `except Exception`. It is like saying "There might be some bug in the code", and we should always try to elliminate all bugs. Is it possible to try to catch possible errors in places they should raise? What are the exact places an error might occure? Maybe to decrease the try block to only those places?

I don't say we can't do with this, just saying we should at least try to elliminate it.
If it is not possible - move the try block to where the function is being called, as a lot of code in try block is not really good.

Reply:
Specifically here since it is being run at a highly sensitive part of robustas launch the safest thing is to swallow any exception. Since this is not operation critical code and robusta should be able to continue to fully function if there is any exceptions that do occur here it is future-error-proofing this code.

Reply:
Agree, then let's just move this try block to `discover_cluster_provider`, where we call this function